Refactor Dockerfile and pre-install rtt_ros2_integration and rtt_ros2_common_interfaces

meyerj opened the pull request:
This is an alternative to #2. The goal is to provide a Docker image with [rtt_ros2_integration](https://github.com/orocos/rtt_ros2_integration) and [rtt_ros2_common_interfaces](https://github.com/orocos/rtt_ros2_common_interfaces) preinstalled in an overlay workspace.

- Instead of adding a new `Dockerfile` in `ros2_integration/ubuntu` as suggested in #2, the existing `Dockerfile` in `ros2/ubuntu` is modified. At the end both images are about Orocos/ROS integration.
- URLs and branches of source repos are listed in a [.repo file](https://index.ros.org/doc/ros2/Installation/Maintaining-a-Source-Checkout/) and cloned using [vcstool](https://github.com/dirk-thomas/vcstool).
- Dependencies are installed using [rosdep](https://wiki.ros.org/rosdep) and do not have to be listed manually in the `Dockerfile` anymore. A custom yaml file `prereqs.yaml` lists Ruby dependencies which are not yet in the official [rosdep database](https://github.com/ros/rosdistro/tree/master/rosdep). Those are for [orogen](https://github.com/orocos-toolchain/orogen) only, which at the moment is not installed in the `ros2` branch of [orocos_toolchain](https://github.com/orocos-toolchain/orocos_toolchain) when using cmake/colcon. This approach has been taken from [the ros2/nightly builds at osrf/docker_images](https://github.com/osrf/docker_images/blob/master/ros2/nightly/nightly/prereqs.yaml).
- By stopping at target `orocos_toolchain` in a [multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) (`docker build --target orocos_toolchain [...]`), one can still create an image with only the core toolchain packages preinstalled, without the overlay workspace.
- Drop the installation with `DESTDIR` at workspace level. For installing individual packages `DESTDIR` works fine, but at workspace level it requires "hacks" with potential side effects, like setting the `CMAKE_FIND_ROOT_PATH`, `PYTHONPATH`, `PKG_CONFIG_PATH` variables to the staged installation directory, such that packages can find the artifacts of others that have been built before, but not yet installed to the final destination. The downside is that almost all commands to clone, install new system dependencies, build and install a workspace have to be combined in a single `RUN` command, such that the resulting new layer only contains the new dependencies and the installed artifacts, but not sources or build artifacts.
- The installation prefix was changed to `/opt/orocos/${ROS_DISTRO}`. Unfortunately installing packages to `/opt/ros/${ROS_DISTRO}` with colcon is not officially supported and it has negative side effects (https://answers.ros.org/question/314019/how-to-install-colcon-packages-into-system-workspace/). Doing so overwrites the setup files generated by [ament_cmake](https://github.com/ament/ament_cmake).
- A new `orocos_entrypoint.sh` script takes care of sourcing both workspaces, `/opt/ros/${ROS_DISTRO}` and `/opt/orocos/${ROS_DISTRO}` and also appends the necessary commands to `~/.bashrc`. The latter is required for bash-specific extensions to work in the docker container, e.g. tab completion for the `ros2` command. Sourcing the `.bash` setup files in the entrypoint itself, before `bash` is executed as a new process, is useless (as done in [osrf/docker_images](https://github.com/osrf/docker_images/blob/5ee4389841e72474698cd2dce7378b54e3717459/ros/eloquent/ubuntu/bionic/ros-core/ros_entrypoint.sh)).
- Added `--parallel-workers 2` because building typekits and transport libraries is quite memory and CPU intensive and freezes the system with the default number of build jobs, i.e. the number of CPU cores.

I am not sure yet whether the final image (up to stage `overlay`) works for CI of [rtt_ros2_integration](https://github.com/orocos/rtt_ros2_integration) and [rtt_ros2_common_interfaces](https://github.com/orocos/rtt_ros2_common_interfaces) itself, if `/opt/orocos/${ROS_DISTRO}` already has installed versions of the same packages. There could be side effects due to mixing of different versions of the packages, e.g. when testing pull requests. I think it is necessary to provide two versions of the image at https://hub.docker.com/repository/docker/orocos/ros2 for each ROS distro, one with `rtt_ros2_integration` preinstalled and one which stops at stage `orocos_toolchain`. I will test it with automated builds now.

If the approach taken here is accepted, it should also be applied to #1 for ROS 1 images.

meyerj committed:
ros2/ubuntu: refactor Dockerfile and pre-install rtt_ros2_integration and rtt_ros2_common_interfaces

meyerj committed:
ros2/ubuntu: update build hook for Docker Hub to build the image in two stages

The first image stops at stage orocos_toolchain and is pushed to orocos/ros2-ci:${DOCKER_TAG}.
Only the second build command adds the final stage, reusing the cached layers of the first build.

spd-intermodalics commented on ros2/ubuntu/Dockerfile:
```diff
@@ -1,44 +1,85 @@
+        --event-handlers console_direct+ \
+        --install-base /opt/orocos/${ROS_DISTRO} \
+        --merge-install \
+        --parallel-workers 2 \
```
My machine with 16 GB of memory still fails in these conditions:
```suggestion
        --parallel-workers 1 \
```
But Docker Hub image builder may have more resources.

meyerj replied:
I think Docker Hub will not work even with `--parallel-workers 1` because the workers are quite limited in resources according to [this post](https://success.docker.com/article/what-are-the-current-resource-limits-placed-on-automated-builds):

> The current limits placed on Automated Builds are:
> 
> 2 hours
> 2 GB RAM
> 1 CPU
> 30 GB Disk Space

So probably we have to switch to Travis or manual builds. Travis [today](https://docs.travis-ci.com/user/reference/overview/) offers 7.5 GB of memory and 2 cores for Ubuntu builders, which should be enough.

meyerj replied:
Removed in https://github.com/orocos/orocos-docker-images/pull/3/commits/dd58443825930692ae6f3b714f7de5d50c0fed0f.

spd-intermodalics commented on ros2/ubuntu/Dockerfile:
```diff
@@ -1,44 +1,85 @@
+    # Install dependencies with rosdep
+    && apt-get update \
+    && apt-get install -y ruby \
+    && rosdep install -y --from-path . --ignore-src \
```
This call in Ubuntu `focal` is still trying to install `ruby-facets`, maybe by one of its dependees `facets`.

meyerj replied:
Fixed in a6bb359: use `gem` instead.
Debian/Ubuntu package [ruby-facets](https://packages.ubuntu.com/ruby-facets) as tracked by the [rosdep database](https://github.com/ros/rosdistro/blob/edd3289d06a3752bd8c988ae71091b490b0f160d/rosdep/ruby.yaml#L14) has been removed for some reason.

spd-intermodalics commented on ros2/ubuntu/overlay.repos:
```diff
@@ -0,0 +1,9 @@
+  rtt_ros2_integration:
+    type: git
+    url: https://github.com/orocos/rtt_ros2_integration.git
+    version: master
```
The current `master` has a bug in the `CMakeLists.txt` of the `tests` subfolder. So this requires to attend first the fix that I submitted [this PR](https://github.com/orocos/rtt_ros2_integration/pull/20).

meyerj replied:
I merged https://github.com/orocos/rtt_ros2_integration/pull/20 after having applied and tested the alternative solution as discussed today.

spd-intermodalics reviewed (changes requested):
This PR outdates my [previous PR](https://github.com/orocos/orocos-docker-images/pull/2) on preparing a Docker image with ROS2 + Orocos toolchain + Ros 2 integration of Orocos, including `rtt_ros2_common_interfaces`.

Both PRs are limited by the amount of resources needed to compile the typekits in `rtt_ros2_common_interfaces`, but it does the work. Furthermore, this PR is a more clean version of the build process compared to the competing PR, it solves better the Python issues generated there.

So, I would go for this one, but first, in the current form it doesn't compile for `foxy`, since `ruby-facets` are unsuccessfully attempted for installation by `rosdep`. After that small fix, it will be all good to go.

meyerj committed:
ros2/ubuntu: use facets from rubygems.org in Ubuntu Focal

The apt package ruby-facets does not exist anymore.

meyerj committed:
ros2/ubuntu: fix key facets in prereqs.yaml

Defining the key here overrides it for all other Ubuntu distros, too.
So the override needs to have a default and only override it for focal.

meyerj committed:
ros2/ubuntu: remove colcon build --parallel-workers argument

If needed, the number of workers can still be limited using

    docker build [...] --cpuset-cpus 0-1

to limit the build containers to only CPU 0 and 1.

nproc would return 2 in this case and the implicit default of colcon/catkin_tools/catkin_make_isolated
is to build with two CPUs then, even if the build host has more than 2 CPUs.

meyerj committed:
ros2/ubuntu: set MAKEFLAGS="-j1" when building stage overlay on DockerHub

meyerj committed:
ros2/ubuntu: build with a single executor (--executor sequential) and a single make job (-j1) for Docker Hub automated builds

spd-intermodalics reviewed (approved):
Thanks for the alternative to PR #2.
LGTM

It checks OK for `dashing`, `eloquent`, and `foxy`.

spd-intermodalics commented:
The checks at Docker Hub CI fail only because of lack or resources, so the fail cross doesn't mean much.

meyerj commented:
I am going to merge this patch. Automated builds still fail, likely because of resource constraints, but we can build and push images manually and eventually switch to Travis or GitHub actions instead in a follow-up.
